### PR TITLE
test(transformers): add source-load parity coverage

### PR DESCRIPTION
**Pull request by yuhezhang-ai**
# What does this PR do ?

Adds source-load parity coverage for NeMo force-HF/HF source loads. The checkpoint-robustness harness now captures logits from the raw HF source load, releases that HF model, constructs the normal AutoModel trainer model, and compares the constructed pre-training model against the HF reference before training starts.

This is stacked on #2942 / `yuhez/fix/tie-weights-nemo-config-gate` and is meant to catch bugs where AutoModel construction silently changes the loaded HF model state before training begins, such as the tied/untied `lm_head` issue fixed there.

# Changelog

- Add opt-in checkpoint-robustness Phase 0 source-load parity checks:
  - raw HF source-load reference logits
  - constructed trainer logits before training
  - max KL, mean KL, and cosine thresholds
  - optional `lm_head`/embedding alias check when storage is inspectable
- Enable real LLM recipe coverage for:
  - `examples/llm_finetune/nemotron/nemotron_nano_9b_squad_peft.yaml`
  - `examples/llm_finetune/llama3_2/llama3_2_1b_hellaswag.yaml`
  - `examples/llm_finetune/ernie4_5/ernie4_5_21b_a3b_hellaswag.yaml`
  - `examples/llm_finetune/nemotron/nemotron_nano_8b_v1_squad.yaml` as allow-failure, tracked by AM-647
  - `examples/llm_finetune/nemotron/nemotron_nano_8b_v1_squad_peft.yaml` as allow-failure, tracked by AM-647
  - `examples/llm_finetune/qwen/qwen3_moe_30b_hellaswag.yaml` as allow-failure, tracked by AM-648
  - `examples/llm_finetune/mistral/ministral3_3b_squad.yaml` as allow-failure, tracked by AM-154
  - `examples/llm_finetune/mistral/ministral3_3b_squad_peft.yaml` as allow-failure, tracked by AM-154
  - `examples/llm_finetune/mistral/mistral_7b_squad.yaml`
  - `examples/llm_finetune/mistral/mistral_7b_hellaswag_fp8.yaml` with an FP8-only mean-KL threshold exception
  - `examples/llm_finetune/mistral/mistral_nemo_2407_squad.yaml`
  - `examples/llm_finetune/mistral/mixtral-8x7b-v0-1_squad.yaml`
- Add source-load parity knobs to Nano 4B and regular Nano 9B too; those generated jobs remain gated by their existing known-issue markers.
- Add CPU proxy unit tests for remote-code list-form tied-key loads and native Mistral causal LM loads, covering both tied and untied embeddings.
- Add CI config resolver support and docs for the new parity knobs.
- Harden Phase 0 for sharded trainers by skipping uninspectable sharded-storage alias checks and using launch-scoped pre-init sync markers.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed Contributor guidelines
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

# Additional Information

- Stacked on #2942.
- Final scoped CI for broadened real-recipe coverage:
  - Parent pipeline: [57179730](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57179730)
  - Generated LLM pipeline: [57179866](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57179866)
  - Blocking jobs passed: ERNIE 4.5 21B A3B, Llama3.2 1B, Nano9 PEFT.
  - Known-issue allow-failure jobs fired as expected: Nano8 regular/PEFT (AM-647), Qwen3 MoE (AM-648).
- Ministral3 add-back scoped CI:
  - Parent pipeline: [57185572](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57185572)
  - Generated LLM pipeline: [57185937](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57185937)
  - Both jobs reached Phase 0 and failed in raw HF source-load forward because Transformers finegrained-FP8 requires the missing `kernels` package; this is expected and non-blocking under AM-154.
- Native Mistral/Mixtral scoped CI:
  - Parent pipeline: [57202130](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57202130)
  - Generated LLM pipeline: [57202390](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57202390)
  - `mistral_7b_squad` passed.
  - `mistral_nemo_2407_squad` reached Phase 0 and exposed a source-load parity mismatch: KL `2.452147e+00`, cosine `0.94497550` (AM-650).
  - `mixtral-8x7b-v0-1_squad` did not reach numeric parity; it failed in the existing robustness setup with `pp_batch_size 2 // pp_microbatch_size 1 must be >= pp_size 8`.
  - Status: failed, kept intentionally as coverage/discussion signal for native Mistral/Mixtral rather than reverted.
- ERNIE threshold tuning scoped CI:
  - Parent pipeline: [57279842](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57279842)
  - Generated LLM pipeline: [57280206](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57280206)
  - `ernie4_5_21b_a3b_hellaswag` passed with Phase 0 `source_load_kl_threshold: 5e-3` and Phase 4 `hf_kl_threshold: 1.1e-2`.
- Max/mean KL gate calibration scoped CI:
  - Three-job parent pipeline: [57417377](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57417377), generated LLM pipeline: [57440557](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57440557)
  - Supplemental parent pipeline: [57439789](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57439789), generated LLM pipeline: [57440080](https://gitlab-master.nvidia.com/dl/JoC/nemo-ci/-/pipelines/57440080)
  - Nano9 PEFT, Llama3.2 1B, and plain Mistral 7B passed the strict default `source_load_mean_kl_threshold: 1e-3`.
  - Mistral 7B FP8 measured max KL `6.488060e-03`, mean KL `2.014334e-03`, cosine `0.99942321`; the recipe therefore uses an FP8-only `source_load_mean_kl_threshold: 3e-3` while non-quantized Phase 0 configs keep the strict default.
  - Gemma3 270M measured small but broad source-load drift (max KL `6.914885e-03`, mean KL `3.045701e-03`, cosine `0.99959433`) and was removed from this PR's Phase 0 coverage because it is legacy/deprecating; Gemma4 belongs in the VLM follow-up.
- Newly exposed issues:
  - AM-647: Nano8 source-load parity mismatch, KL `7.357722e+00`, cosine `0.92351776`.
  - AM-648: Qwen3 MoE source-load parity mismatch, KL `1.815263e+01`, cosine `0.81525248`.
  - AM-650: Mistral-Nemo 2407 source-load parity mismatch, KL `2.452147e+00`, cosine `0.94497550`.
- Coverage notes:
  - This PR uses representative coverage by load path rather than enabling Phase 0 on every checkpoint-robustness recipe. Many robustness recipes are PEFT/chat/packing variants over the same base checkpoint, so they would repeat the same source-load signal while increasing scheduled CI time.
  - The current real-config set covers remote-code/force-HF Nemotron Nano, tied controls, native Llama/Mistral, FP8 Mistral, MoE/adaptor-style models, and known failing paths kept visible as allow-failure.
  - ERNIE keeps Phase 0 source-load parity strict: the rerun measured KL `8.728858e-04`, cosine `0.99961627`, so `source_load_kl_threshold` is `5e-3`. The separate Phase 4 HF-reload check is the known AM-428 family; the same rerun measured `HF-loaded max KL=1.006250e-02`, so ERNIE uses a narrow `hf_kl_threshold: 1.1e-2` instead of the earlier temporary `2e-2`.
  - Native Mistral/Mixtral coverage is intentionally present now; Mistral-Nemo 2407 exposes a real mismatch and Mixtral exposes a robustness harness/config blocker, so follow-up policy can decide whether to track, allow-fail, or drop these recipes if they are being deprecated.
  - Ministral3 source-load parity is enabled now, even though the current raw HF path is blocked by AM-154 before numeric metrics.
  - Gemma4 should be tracked separately through VLM/model enablement; the current checkpoint-robustness harness here is LLM-only.
- Validation on login node used `source work/runs/_shared/env.sh` per `work/agents.md`:
  - `BASH_ENV=work/runs/_shared/env.sh /usr/bin/bash -c 'uv run --no-sync pytest tests/unit_tests/ci_tests/test_config_resolver.py -q'` (rerun after final Gemma3/FP8 config calibration)
  - `work/runs/.venvs/Automodel/bin/python -m pytest tests/unit_tests/ci_tests/test_config_resolver.py -q` (rerun after ERNIE threshold tuning)
  - `work/runs/.venvs/Automodel/bin/python -m pytest tests/unit_tests/_transformers/test_model_init.py::TestTieWeightsNemoConfigGate tests/unit_tests/_transformers/test_source_load_parity.py -q -p no:cacheprovider`
  - `work/runs/.venvs/Automodel/bin/python -m py_compile tests/functional_tests/checkpoint_robustness/test_checkpoint_robustness_llm.py`
  - `ruff check tests/functional_tests/checkpoint_robustness/test_checkpoint_robustness_llm.py tests/unit_tests/_transformers/test_source_load_parity.py`
  - `ruff format tests/functional_tests/checkpoint_robustness/test_checkpoint_robustness_llm.py tests/unit_tests/_transformers/test_source_load_parity.py --check`

### CODE REVIEW

**Comment by yuhezhang-ai** [comment]
/ok to test [f546c1e](https://github.com/NVIDIA-NeMo/Automodel/pull/2960/commits/f546c1ede1ffc5c402afdb9fb6e0ade69127ccdc)

**Comment by yuhezhang-ai** [comment]
/ok to test [ba22807](https://github.com/NVIDIA-NeMo/Automodel/pull/2960/commits/ba22807bb98b4a4cfbbf3af30dbf9916b8f1b1a9)

File: examples/llm_finetune/nemotron/nemotron_nano_4b_squad.yaml [hunk 1]
```diff
@@ -94,4 +94,17 @@ lr_scheduler:
...
 ci:
   recipe_owner: adil-a
+  time: "00:20:00"
   known_issue_id: AM-153
+  checkpoint_robustness:
+    check_source_load_parity: true
+    kl_threshold: 5e-3
+    hf_kl_threshold: 5e-3
+    source_load_kl_threshold: 5e-3
+    source_load_cosine_threshold: 0.9999
```
Comment (akoumpa) [inline examples/llm_finetune/nemotron/nemotron_nano_4b_squad.yaml]: @yuhezhang-ai others have 0.9995 vs this one 0.9999; I wonder if we could use 0.9999 in all.

**Review by akoumpa — DISMISSED** [review]

File: examples/llm_finetune/nemotron/nemotron_nano_4b_squad.yaml (hunk 1 above)
Comment (yuhezhang-ai) [reply examples/llm_finetune/nemotron/nemotron_nano_4b_squad.yaml]: Good catch. Nano 4B had 0.9999 only because it inherited the default; we do not have evidence it should be stricter because AM-153 blocks it before Phase 0. Nano 9B already showed 0.9999 is too tight for this bf16 source-load path, so I aligned Nano 4B with the other non-FP8 configs at 0.9995.

**Comment by yuhezhang-ai** [comment]
/ok to test [9b362a3](https://github.com/NVIDIA-NeMo/Automodel/pull/2960/commits/9b362a33cf1b70b17895fa19edf384dc9c88da0b)

**Review by akoumpa — APPROVED** [review]

**Comment by yuhezhang-ai** [comment]
/ok to test [00b5ac3](https://github.com/NVIDIA-NeMo/Automodel/pull/2960/commits/00b5ac33a76d72c6bfa0ef4e5b71747a94435e35)

**Review by thomasdhc — APPROVED** [review]
